### ISSUE TITLE
Celery 5 hangs when tasks emit lots of logs

### ISSUE THREAD

**Issue by lelit** [celery/celery#6372]
I tried switching my application to Celery 5, and running the test suite result in a complete hang, in a particular test. The suite uses a worker running with `solo` pooling and `concurrency=1`, but trying to switch to a more standard configuration didn't make any difference. While testing, the logging of the whole worker and all the tasks is configured at `DEBUG` level, so it is considerably noisy.

As there's nothing special happening there, I tried to investigate further:

- executing that single test, the hang does **not** happen
- executing just a few tests before that up to the offending one, the hang does **not** happen
- putting a `pdb` in the offending test, and just keep hitting `n` to advance, the hang does **not** happen
- wrapping the worker within [snaketrace](https://pypi.org/project/snaketrace/), the hang happens almost immediately, at the very first test in the suite, one that basically does nothing except testing the serialization mechanism asserting that a `date` value passed as argument effectively lands as a native `date`
- redirecting the output of `snaketrace` to a file, the hang is postponed again at the previous test

The `snaketrace` output does not seem to reveal anything useful: at one point, everything stops; when I interrupt the test runner (`pytest`), I see that the offending test successfully completes its work, and eventually the worker shuts down as usual.

The application is running in production since a couple of years now, with Celery 4.4.7, and obviously the whole test suite completes without error in that context, and I never experienced any kind of hang.

I'm not (yet?) able to setup a minimal example that reproduces the issue, but I will keep trying.

$ celery -A etour.hopi.atc report

software -> celery:5.0.0 (singularity) kombu:5.0.2 py:3.8.5
            billiard:3.6.3.0 redis:3.5.3
platform -> system:Linux arch:64bit
            kernel version:5.6.18 imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:redis://redis:6379/0

broker_url: 'redis://redis:6379/0'
result_backend: 'redis://redis:6379/0'
include: ['etour.hopi.celerer.tasks',
 'etour.hopi.renderer.tasks',
 'etour.hopi.slurper.tasks']
deprecated_settings: None

**Comment by thedrow** [comment]
I don't think anything changed in that area of the code so I'm not sure what could be the root cause of this issue.
Please let me know if you have a test case available.

On another matter, we have a bug report template which we require our users to use. It eventually pays off.
I'm going to keep this issue opened simply because it's descriptive enough but please refrain from opening a bug report without using our template.

**Comment by lelit** [comment]
First of all a _brown paper bag_ omission on my part: the app in question is composed of several components, each of them running inside a Docker container where one of them is the Celery `worker` process, another the `beat` one, and several others that executes async tasks. 

The test suite exercises them all, one after the other: for some reason the test runner restarts the `worker` container for each component, that is, the suite for each component is wrapped between a `docker start celery_worker` and a `docker stop celery_worker`. It does not seem an hard requirement, but it has been working like that since its inception, with Celery 4.3.x and Celery 4.4.x. 

After a considerable amount of trials, a few days ago I found the *thing* that apparently causes/cures my problem: changing the way the test suites are executed, i.e. commenting out the statements that forcefully stop the `worker` container after each suite, the hang I was experiencing with Celery 5 disappeared! Restoring them, the hang invariably happens again.

With that knowledge, I planned to spend some more time on my attempt to produce a minimal example that replicates the problem. I was about to mimic the above logic, but having noticed the release of Celery 5.0.1 I decided to try that first and... it **just works**! I mean, obliterating the tiny changes I made to the test runner, after repeated runs I never experienced the hang again.

 Although I feel a bit of pain due to my inability to reproduce and/or isolate what is causing the hang (also, browsing the commits between 5.0.0 and 5.0.1 didn't ring any bell to me), I'm obviously very happy!
